### ISSUE TITLE
Reg loss don't decrease after epoch 100, could you give me some suggestions?

### ISSUE THREAD

**Issue by wulongjian** [zylo117/Yet-Another-EfficientDet-Pytorch#463]
Hello, First thanks for your code. 
When I train my custom dataset with your pretrained model edet-0 after epoch 100, the Cls loss reduce from 1000 to 0.5 nearly, but the Reg loss almost don't change, about 2.5. I have  also view other issues, it seems others often meet cls loss much big, not the Reg loss. I'm wondering if I need to change something like anchor_size, I follow this [mnslarcher/kmeans-anchors-ratios](https://github.com/mnslarcher/kmeans-anchors-ratios) just change the ahchors_ratios, should I change the ahchors_scales or something else?
Here is my run code and result
`python train.py -c 0 -p train_spring --batch_size 16 --lr 2e-4 --load_weights logs-head/train_spring/efficientdet-d0.pth --head_only True --saved_path logs-head/ --num_epochs 100`
`Step: 30575. Epoch: 97/100. Iteration: 312/312. Cls loss: 0.61165. Reg loss: 2.73616. Total loss: 3.34781: 100%|███████████████████████████████████| 312/312 [03:12<00:00,  1.62it/s]

Val. Epoch: 97/100. Classification loss: 0.53079. Regression loss: 2.81285. Total loss: 3.34364
Step: 30887. 
Epoch: 98/100. Iteration: 312/312. Cls loss: 0.42236. Reg loss: 2.20621. Total loss: 2.62857: 100%|███████████████████████████████████| 312/312 [03:34<00:00,  1.45it/s]

Val. Epoch: 98/100. Classification loss: 0.53064. Regression loss: 2.81417. Total loss: 3.34481
Step: 30999. 
Epoch: 99/100. Iteration: 112/312. Cls loss: 0.56077. Reg loss: 2.89549. Total loss: 3.45626:  36%|████████████▍                      | 111/312 [01:22<01:43,  1.94it/s]checkpoint...
Step: 31199. Epoch: 99/100. Iteration: 312/312. Cls loss: 0.49382. Reg loss: 2.51970. Total loss: 3.01351: 100%|███████████████████████████████████| 312/312 [03:18<00:00,  1.58it/s]

Val. Epoch: 99/100. Classification loss: 0.53125. Regression loss: 2.81609. Total loss: 3.34734
`
Thanks for your reply.

**Comment by zylo117** [comment]
Reg loss is 10x higher than expected. What's your anchor configs?

**Comment by wulongjian** [comment]
Here is my anchor configs
```
anchors_scales: '[2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)]'
anchors_ratios: '[(1.0, 1.0), (1.1, 0.9), (1.3, 0.8)]'
```
I follow this tutorial [mnslarcher/kmeans-anchors-ratios](https://github.com/mnslarcher/kmeans-anchors-ratios) change anchor, 
`python kmeans_anchors_ratios.py \
    --instances ../edet/datasets/train_spring/annotations/instances_train2019.json \
    --anchors-sizes 32 64 128 256 512 \
    --input-size 512 \
    --normalizes-bboxes True \
    --num-runs 3 \
    --num-anchors-ratios 3 \
    --max-iter 300 \
    --min-size 0 \
    --iou-threshold 0.5 \
    --decimals 1 \
    --default-anchors-ratios '[(1.0, 1.0), (1.4, 0.7), (0.7, 1.4)]'
`
I just change the json file, I don't change such as --anchor-sizes and other parameters, I think my anchor configs is wrong, but I don't know how to fix it.

**Comment by zylo117** [comment]
Have you trained on other network? Did it work?
If not, then it's probably that your dataset is too badly labeled or it's just too hard to be learned.

**Comment by wulongjian** [comment]
well, I have trained this in ATSS(CVPR 2020), it performs much better than edet, here is the AP results.
```
ATSS:
[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.414
[ IoU=0.50      | area=   all | maxDets=100 ] = 0.766
[ IoU=0.75      | area=   all | maxDets=100 ] = 0.397
EDet:
[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.080
[ IoU=0.50      | area=   all | maxDets=100 ] = 0.190
[ IoU=0.75      | area=   all | maxDets=100 ] = 0.054
```
In ATSS, there exist anchor size config like follows, 
 ```
ATSS:

    ANCHOR_SIZES: (48,96,192,384,768) 

    ASPECT_RATIOS: (1.0,)
```
well, I not familiar with your code, Is there some way to set anchor_size like atss in your code?

**Comment by DavidChou2017** [comment]
hi~  I met the same problem when I tried to train ef-d1 with the custom dataset with the command 'python train.py -c 1 -p abnormal_detection --head_only False --lr 0.0001 --batch_size 8 --num_epochs 200' . My reg loss is about 0.5, while the cls loss is about 3.0.  If you have solved this problem, I appreciate it if you'd like to share your solution. Thanks again.

**Comment by SubramanianKrish** [comment]
Hi! I'm seeing similar results too.. Could you explain any solution you may have come up with? I'm assuming there's a solution as the issue has been closed

**Comment by WangBoying** [comment]
hi! I met the same problem. Can you tell me how you solved it? Thanks！

**Comment by WangBoying** [comment]
@zylo117 @wulongjian

**Comment by zylo117** [comment]
For most cases, it's just underfitting, keep training with larger lr.

**Comment by WangBoying** [comment]
I has used lr = 1e-1 and lr = 1e-2, but the problem still exists.

**Comment by SubramanianKrish** [comment]
Start training with some initial lr (say 1e-2 or 1e-3), train for a few epochs (depends on your dataset). The code's LR scheduler will kick in at some point and reduce the lr to a smaller value. Say you train for 100 epochs, the lr has come down to 1e-5 when started with 1e-2, stop training. Take the best weights from that experiment and using that as the initial weights, train another 50 epochs (these numbers are just examples) with starting lr 1e-3. See if the loss explodes.. if it does start with a lower lr. In most cases (at least in my experiments), retraining using best weights of an experiment but starting at a higher lr helped!

P.S: (Untested) Maybe try playing with the patience factor in LR scheduler. This may help to keep the learning rate higher for longer periods of times.

**Comment by zylo117** [comment]
I don't think 1e-1 or 1e-2 is a reasonable starting point. It's too high.

**Comment by 0K2333** [comment]
@wulongjian 我遇到了和你完全一样的问题！！请问您最后是如何解决的，可以调整anchor_size吗？
> 好吧，我已经在 ATSS（CVPR 2020） 中训练了它，它的性能比 edet 好得多，这是 AP 结果。
> 
> ```
> ATSS:
> [ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.414
> [ IoU=0.50      | area=   all | maxDets=100 ] = 0.766
> [ IoU=0.75      | area=   all | maxDets=100 ] = 0.397
> EDet:
> [ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.080
> [ IoU=0.50      | area=   all | maxDets=100 ] = 0.190
> [ IoU=0.75      | area=   all | maxDets=100 ] = 0.054
> ```
> 
> 在 ATSS 中，存在如下锚点大小配置，
> 
> ```
> ATSS:
> 
>    ANCHOR_SIZES: (48,96,192,384,768) 
> 
>    ASPECT_RATIOS: (1.0,)
> ```
